File: neetcode/java/src/main/java/solutions/arraysandhashing/TwoSum.java
package solutions.arraysandhashing;

import java.util.HashMap;
/*
 * https://leetcode.com/problems/two-sum/
 *
 * Explanation: If the other operand is not found, we add the current operand 
 *  to the map. That way, when the other operand is reached, it can find the 
 *  operand that previously conducted the search. This prevents repeat 
 *  solutions from being considered as the same.
 *
 * Complexity
 *  Time:   O(n) - if no target is found the entire nums array is incremented 
 *  Space:  O(n) - if no target is found, a map of size n is filled
 */
public class TwoSum{
  public static int[] twoSum(int[] nums, int target){
    HashMap<Integer, Integer> map = new HashMap<>();
    for (int i = 0; i < nums.length; i++){
      int diff = target - nums[i];
      if (map.containsKey(diff))
        return new int[]{map.get(diff), i};
      else
        map.put(nums[i], i);
    }
    return null;
  }
}

File: neetcode/python/src/arraysandhashing/two_sum.py
from typing import List

class Solution:
    def twoSumNaive(nums: List[int], target: int) -> List[int]:
        for i in range(0, len(nums)):
            for j in range(i+1, len(nums)):
                if nums[i] + nums[j] == target:
                    return [i,j]
        return [0,0]

    def twoSum(nums: List[int], target: int) -> List[int]:
        dictNums = dict()
        for i in range(len(nums)):
            search = target - nums[i]
            if search in dictNums:
                return [dictNums[search], i]
            dictNums[nums[i]] = i
        return [-1,-1]
